File: LeetCodeTasks/24_swap-nodes-in-pairs/solutions/python3/solution.py
from typing import Optional



class Solution:

    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:

        dummy = ListNode(0)

        dummy.next = head

        prev = dummy

        

        while head and head.next:

            first = head

            second = head.next

            

            # Swapping

            prev.next = second

            first.next = second.next

            second.next = first

            

            # Reinitializing the pointers for next swap

            prev = first

            head = first.next

            

        return dummy.next

File: LeetCodeTasks/24_swap-nodes-in-pairs/solutions/cpp/solution.cpp
#include <iostream>


class Solution {

public:

    ListNode* swapPairs(ListNode* head) {

        if (head == nullptr || head->next == nullptr) return head;

        

        ListNode *dummy = new ListNode(0);

        dummy->next = head;

        ListNode *current = dummy;

        

        while (current->next != nullptr && current->next->next != nullptr) {

            ListNode *first = current->next;

            ListNode *second = first->next;

            

            first->next = second->next;

            second->next = first;

            current->next = second;

            

            current = first;

        }

        

        ListNode *newHead = dummy->next;

        delete dummy;

        return newHead;

    }

};

